### PR TITLE
Moved and tidied Cnvnator, created tests and modified mip.pl accordingly.

### PR DESCRIPTION
**Pull request by northwestwitch**
Tested on rasta, it works fine!

### CODE REVIEW

File: lib/MIP/Program/Variantcalling/Cnvnator.pm [hunk 1]
```diff
@@ -0,0 +1,561 @@
...
+
+## Function : Perl wrapper for writing cnvnator recipe to $FILEHANDLE or return commands array. Based on cnvnator 0.3.3.
+## Returns  : "@commands"
+## Arguments: $regions_ref, $infile_path, $stderrfile_path, $stdoutfile_path, $FILEHANDLE, $cnv_bin_size
+##          : $regions_ref     => The regions to process {REF}
+##          : $infile_path     => Infile paths
+##          : $stderrfile_path => Stderrfile path
+##          : $stdoutfile_path => Stdoutfile path
+##          : $FILEHANDLE      => Filehandle to write to
+##           : $cnv_bin_size    => Copy number variant bin size
```
Comment (henrikstranneheim) [inline lib/MIP/Program/Variantcalling/Cnvnator.pm]: align

File: lib/MIP/Program/Variantcalling/Cnvnator.pm [hunk 2]
```diff
@@ -0,0 +1,561 @@
...
+
+## cnvnator_calling
+
+## Function : Perl wrapper for writing cnvnator recipe to $FILEHANDLE or return commands array. Based on cnvnator 0.3.3.
+## Returns  : "@commands"
+## Arguments: $regions_ref, $infile_path, $outfile_path, $stderrfile_path, $stdoutfile_path, $FILEHANDLE, $cnv_bin_size
+##          : $regions_ref             => The regions to process {REF}
+##          : $infile_path             => Infile paths
+##          : $outfile_path            => Outfile path
+##          : $stderrfile_path         => Stderrfile path
```
Comment (henrikstranneheim) [inline lib/MIP/Program/Variantcalling/Cnvnator.pm]: Align fat commas

File: lib/MIP/Program/Variantcalling/Cnvnator.pm [hunk 3]
```diff
@@ -0,0 +1,561 @@
...
+
+        push @commands, q{-call} . $SPACE . $cnv_bin_size;
+    }
+
+    ## Infile
+    push @commands, q{-root}, $infile_path;
+
+    if ($outfile_path) {
+
+        #Specify output filename
```
Comment (henrikstranneheim) [inline lib/MIP/Program/Variantcalling/Cnvnator.pm]: #Specify -> #Specify

File: lib/MIP/Program/Variantcalling/Cnvnator.pm [hunk 4]
```diff
@@ -0,0 +1,561 @@
...
+    }
+
+    if ($unique) {
+
+        push @commands, q{-unique};
+    }
+
+    if ($outfile_path) {
+
+        #Specify output filename
```
Comment (henrikstranneheim) [inline lib/MIP/Program/Variantcalling/Cnvnator.pm]: #Specify -> #Specify

File: lib/MIP/Program/Variantcalling/Cnvnator.pm [hunk 5]
```diff
@@ -0,0 +1,561 @@
...
+
+    if ($cnv_bin_size) {
+
+        push @commands, q{-call} . $SPACE . $cnv_bin_size;
+    }
+
+    ## Infile
+    push @commands, q{-root}, $infile_path;
+
+    if ($outfile_path) {
```
Comment (henrikstranneheim) [inline lib/MIP/Program/Variantcalling/Cnvnator.pm]: Is this not replaced by stdoutfile_path?

File: lib/MIP/Program/Variantcalling/Cnvnator.pm [hunk 6]
```diff
@@ -0,0 +1,561 @@
...
+    ## Infile
+    push @commands, $infile_path;
+
+    ## Options
+    if ($referencedirectory_path) {
+
+        push @commands, $referencedirectory_path;
+    }
+
+    if ($outfile_path) {
```
Comment (henrikstranneheim) [inline lib/MIP/Program/Variantcalling/Cnvnator.pm]: Replace by stdoutfile_path?

File: t/cnvnator_histogram.t [hunk 7]
```diff
@@ -0,0 +1,197 @@
...
+    },
+    referencedirectory_path => {
+        input           => q{path_to_ref_directory},
+        expected_output => q{-d path_to_ref_directory},
+    },
+);
+
+my %specific_argument = (
+    regions_ref => {
+        inputs_ref => [qw{ -chrom 12:11396601-11436500 14:12000000-13000000}],
```
Comment (henrikstranneheim) [inline t/cnvnator_histogram.t]: should you not remove -chrom from the input?

File: t/cnvnator_partition.t [hunk 8]
```diff
@@ -0,0 +1,193 @@
...
+my %required_argument = (
+    infile_path => {
+        input           => q{infile_path},
+        expected_output => q{infile_path},
+    },
+);
+
+my %specific_argument = (
+    regions_ref => {
+        inputs_ref => [qw{ -chrom 12:11396601-11436500 14:12000000-13000000}],
```
Comment (henrikstranneheim) [inline t/cnvnator_partition.t]: same here

File: t/cnvnator_read_extraction.t [hunk 9]
```diff
@@ -0,0 +1,195 @@
...
+my %required_argument = (
+    infile_paths_ref => {
+        inputs_ref      => [qw{ file1.bam file2.bam file3.bam }],
+        expected_output => q{--tree file1.bam file2.bam file3.bam},
+    },
+);
+
+my %specific_argument = (
+    regions_ref => {
+        inputs_ref => [qw{ -chrom 12:11396601-11436500 14:12000000-13000000}],
```
Comment (henrikstranneheim) [inline t/cnvnator_read_extraction.t]: and here

File: t/cnvnator_statistics.t [hunk 10]
```diff
@@ -0,0 +1,193 @@
...
+my %required_argument = (
+    infile_path => {
+        input           => q{infile_path},
+        expected_output => q{infile_path},
+    },
+);
+
+my %specific_argument = (
+    regions_ref => {
+        inputs_ref => [qw{ -chrom 12:11396601-11436500 14:12000000-13000000}],
```
Comment (henrikstranneheim) [inline t/cnvnator_statistics.t]: and here

File: lib/MIP/Program/Variantcalling/Cnvnator.pm (hunk 5 above)
Comment (northwestwitch) [reply lib/MIP/Program/Variantcalling/Cnvnator.pm]: They are both there in the original code, then if you have one or the other you push to the commands either:

> . $SPACE . $outfile_path or 
1> . $PACE . $stdoutfile_path

Do you want me to write something like:

if($outfile_path ne q{})
{
$stdoutfile_path = $outfile_path;
}

and then I pass $stdoutfile_path to the standard streams?

Or you had something else in mind?

File: t/cnvnator_histogram.t (hunk 7 above)
Comment (northwestwitch) [reply t/cnvnator_histogram.t]: Believe me I tried, but this is the only thing that works for me to make it pass the test. I've tried with single chromosomes like 1 2 3, or chr1 chr2 chr3 but also regions, like 1:290910-8198911 or chr1:9019201-89898989. It all fails!

File: lib/MIP/Program/Variantcalling/Cnvnator.pm (hunk 5 above)
Comment (henrikstranneheim) [reply lib/MIP/Program/Variantcalling/Cnvnator.pm]: just use $stdoutfile_path

File: t/cnvnator_histogram.t (hunk 7 above)
Comment (henrikstranneheim) [reply t/cnvnator_histogram.t]: strange we can look at it together on monday

**Comment by northwestwitch** [comment]
Yes, sure!

**Comment by northwestwitch** [comment]
Ok, I'll also test it on rasta maybe!

**Comment by henrikstranneheim** [comment]
👍

**Pull request merged**